### PR TITLE
planner: support more types to use IndexMerge to access MVIndex

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve?

Issue Number: ref #40191

Problem Summary: planner: support more types to use IndexMerge to access MVIndex

### What is changed and how it works?

planner: support more types to use IndexMerge to access MVIndex

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by Yisaer — APPROVED** [review]

File: planner/core/indexmerge_path.go [hunk 1]
```diff
@@ -642,16 +628,8 @@ func (ds *DataSource) generateIndexMergeJSONMVIndexPath(normalPathCnt int, filte
...
-	arrayExpr, wrappedByJSONCast := unwrapJSONCast(jsonArrayExpr)
-	if !wrappedByJSONCast {
-		return nil, false
-	}
-	if _, isConst := arrayExpr.(*expression.Constant); !isConst {
-		return nil, false
-	}
-	if expression.IsMutableEffectsExpr(arrayExpr) {
+func jsonArrayExpr2Exprs(sctx sessionctx.Context, jsonArrayExpr expression.Expression, targetType *types.FieldType) ([]expression.Expression, bool) {
+	if !expression.IsInmutableExpr(jsonArrayExpr) || jsonArrayExpr.GetType().EvalType() != types.ETJson {
```
Comment (xiongjiwei) [inline planner/core/indexmerge_path.go:632]: how about use `FoldConst`?

File: planner/core/indexmerge_path.go [hunk 2]
```diff
@@ -677,16 +655,18 @@ func jsonArrayExpr2Exprs(sctx sessionctx.Context, jsonArrayExpr expression.Expre
...
-func jsonValue2Expr(v types.BinaryJSON) (expression.Expression, bool) {
-	if v.TypeCode != types.JSONTypeCodeInt64 {
-		// only support INT now
-		// TODO: support more types
+func jsonValue2Expr(sctx sessionctx.Context, v types.BinaryJSON, targetType *types.FieldType) (expression.Expression, bool) {
+	convertFunc := expression.ConvertJSON2Tp(targetType.EvalType())
+	if convertFunc == nil {
+		return nil, false
+	}
+	datum, err := convertFunc(sctx.GetSessionVars().StmtCtx, v, targetType)
```
Comment (xiongjiwei) [inline planner/core/indexmerge_path.go:663]: we can not use `sctx.GetSessionVars().StmtCtx` directly here. If the sql mode is `null` some error is ignored

**Review by xiongjiwei — APPROVED** [review]
Rest LGTM

File: planner/core/indexmerge_path.go (hunk 2 above)
Comment (xiongjiwei) [reply planner/core/indexmerge_path.go:663]: please ref the usage in expression package

File: planner/core/indexmerge_path.go (hunk 1 above)
Comment (qw4990) [reply planner/core/indexmerge_path.go:642]: No, I've tried this and found that `FoldConst` cannot handle expressions with nested Scalar-Functions like `add(1, add(1, 1))`.

**Pull request merged**